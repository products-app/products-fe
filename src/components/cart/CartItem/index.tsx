import Image from 'next/image'
import { Text } from '@lebernardo/react'
import { Minus, Plus, TrashSimple } from 'phosphor-react'
import { formatDecimalToReal, getProductImage } from '@/helpers/products'
import {
  ButtonQuantityPlus,
  ButtonQuantityMinus,
  ContainerImg,
  ContainerText,
  TextQuantity,
  ContainerControls,
  TextPrice,
  ButtonRemoveProduct,
  ContainerQuantityControl,
} from './styles'

type CartItemProps = {
  item: app.CartItem
  increaseItem: (item: app.CartItem) => void
  decreaseItem: (item: app.CartItem) => void
  removeItem: (item: app.CartItem) => void
}

const CartItem = ({
  item,
  increaseItem,
  decreaseItem,
  removeItem,
}: CartItemProps) => {
  return (
    <>
      <ContainerImg>
        <Image
          src={getProductImage(item.image)}
          alt={item.name}
          width={0}
          height={0}
          sizes="100vw"
        />
      </ContainerImg>

      <ContainerText>
        <Text size="sm">{item.name}</Text>
        <TextQuantity size="xs">
          Qtde:{' '}
          <strong data-cy="item-quantity" aria-label="item-quantity">
            {item.quantity}
          </strong>
        </TextQuantity>
      </ContainerText>

      <ContainerControls>
        <TextPrice>{formatDecimalToReal(item.price * item.quantity)}</TextPrice>
        <ContainerQuantityControl>
          <ButtonQuantityMinus
            data-cy="decrease-item"
            onClick={() => decreaseItem(item)}
            disabled={item.quantity === 1}
          >
            <Minus />
          </ButtonQuantityMinus>
          <ButtonQuantityPlus
            data-cy="increase-item"
            onClick={() => increaseItem(item)}
            disabled={item.quantity === item.stock}
          >
            <Plus />
          </ButtonQuantityPlus>
        </ContainerQuantityControl>
      </ContainerControls>

      <ContainerControls>
        <ButtonRemoveProduct onClick={() => removeItem(item)}>
          <TrashSimple />
        </ButtonRemoveProduct>
      </ContainerControls>
    </>
  )
}

export default CartItem
