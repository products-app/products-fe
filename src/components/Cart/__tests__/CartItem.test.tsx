import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import CartItem from '../CartItem/index'

const props = {
  item: {
    id: 1,
    name: 'Product 1',
    price: 29.9,
    stock: 10,
    quantity: 3,
  },
  increaseItem: jest.fn(),
  decreaseItem: jest.fn(),
  removeItem: jest.fn(),
}

describe('Cart Item component', () => {
  test('should render correctly', () => {
    const { container } = render(<CartItem {...props} />)

    expect(screen.getByText('Product 1')).toBeInTheDocument()
    expect(() => screen.getByText('no exist')).toThrow()
    expect(container).toMatchSnapshot()
  })

  test('should click on increase quantity button', async () => {
    render(<CartItem {...props} />)

    const handleClick = jest.fn()
    const increaseButtonElement = screen.getAllByRole('button')[1]
    increaseButtonElement.addEventListener('click', handleClick, false)

    await userEvent.click(increaseButtonElement)

    expect(handleClick).toHaveBeenCalled()
  })

  test('should click on decrease quantity button', async () => {
    render(<CartItem {...props} />)

    const handleClick = jest.fn()
    const decreaseButtonElement = screen.getAllByRole('button')[0]
    decreaseButtonElement.addEventListener('click', handleClick, false)

    await userEvent.click(decreaseButtonElement)

    expect(handleClick).toHaveBeenCalled()
  })

  test('should remove an item correctly', async () => {
    render(<CartItem {...props} />)

    const handleClick = jest.fn()
    const decreaseButtonElement = screen.getAllByRole('button')[2]
    decreaseButtonElement.addEventListener('click', handleClick, false)

    await userEvent.click(decreaseButtonElement)

    expect(handleClick).toHaveBeenCalled()
  })
})
