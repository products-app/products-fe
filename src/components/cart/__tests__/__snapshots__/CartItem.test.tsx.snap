// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Item component should render correctly 1`] = `
<div>
  <div
    class="c-dZtilx"
  >
    <img
      alt="Product 1"
      data-nimg="1"
      decoding="async"
      height="0"
      loading="lazy"
      sizes="100vw"
      src="/_next/image?url=%2Fproduct-not-found.jpg&w=3840&q=75"
      srcset="/_next/image?url=%2Fproduct-not-found.jpg&w=640&q=75 640w, /_next/image?url=%2Fproduct-not-found.jpg&w=750&q=75 750w, /_next/image?url=%2Fproduct-not-found.jpg&w=828&q=75 828w, /_next/image?url=%2Fproduct-not-found.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fproduct-not-found.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fproduct-not-found.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fproduct-not-found.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fproduct-not-found.jpg&w=3840&q=75 3840w"
      style="color: transparent;"
      width="0"
    />
  </div>
  <figure
    class="c-hBSDYn"
  >
    <p
      class="c-dgMvMW c-dgMvMW-fAKXXr-size-sm"
    >
      Product 1
    </p>
    <p
      class="c-dgMvMW c-fFOkxT c-dgMvMW-cEPjnm-size-xs"
    >
      Qtde:
       
      <strong
        aria-label="item-quantity"
        data-cy="item-quantity"
      >
        3
      </strong>
    </p>
  </figure>
  <div
    class="c-gsXdEb"
  >
    <p
      class="c-dgMvMW c-dnNpSg c-dgMvMW-kWyXas-size-md"
    >
      R$Â 89,70
    </p>
    <div
      class="c-gJlbUq"
    >
      <button
        class="c-gXmMMa"
        data-cy="decrease-item"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="none"
            height="256"
            width="256"
          />
          <line
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
            x1="40"
            x2="216"
            y1="128"
            y2="128"
          />
        </svg>
      </button>
      <button
        class="c-fOZJzt"
        data-cy="increase-item"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="none"
            height="256"
            width="256"
          />
          <line
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
            x1="40"
            x2="216"
            y1="128"
            y2="128"
          />
          <line
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="16"
            x1="128"
            x2="128"
            y1="40"
            y2="216"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    class="c-gsXdEb"
  >
    <button
      class="c-cHEnSC"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <line
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="216"
          x2="40"
          y1="56"
          y2="56"
        />
        <line
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="88"
          x2="168"
          y1="24"
          y2="24"
        />
        <path
          d="M200,56V208a8,8,0,0,1-8,8H64a8,8,0,0,1-8-8V56"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
    </button>
  </div>
</div>
`;
